File: backend/app/schemas/storm.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class StormAdvisoryLinkResponse(BaseModel):
    advisory_number: str | None = None
    issuance: datetime | None = None
    url: str | None = None


class StormSystemResponse(BaseModel):
    id: str
    bin_number: str | None = None
    name: str | None = None
    basin: str | None = None
    classification: str | None = None
    classification_label: str | None = None
    intensity_knots: float | None = None
    pressure_mb: float | None = None
    latitude: str | None = None
    longitude: str | None = None
    latitude_numeric: float | None = None
    longitude_numeric: float | None = None
    movement_direction_deg: float | None = None
    movement_speed_mph: float | None = None
    last_update: datetime | None = None
    public_advisory: StormAdvisoryLinkResponse | None = None
    forecast_advisory: StormAdvisoryLinkResponse | None = None
    forecast_discussion: StormAdvisoryLinkResponse | None = None
    forecast_graphics: StormAdvisoryLinkResponse | None = None
    forecast_track_kmz_url: str | None = None
    wind_speed_probabilities_url: str | None = None


class StormTrackingResponse(BaseModel):
    source_url: str
    status: Literal["available", "unavailable"]
    fetched_at: datetime | None = None
    message: str | None = None
    active_storms: list[StormSystemResponse] = Field(default_factory=list)

File: frontend/src/types/storm.ts
export interface StormAdvisoryLink {
  advisory_number?: string | null;
  issuance?: string | null;
  url?: string | null;
}

export interface StormSystem {
  id: string;
  bin_number?: string | null;
  name?: string | null;
  basin?: string | null;
  classification?: string | null;
  classification_label?: string | null;
  intensity_knots?: number | null;
  pressure_mb?: number | null;
  latitude?: string | null;
  longitude?: string | null;
  latitude_numeric?: number | null;
  longitude_numeric?: number | null;
  movement_direction_deg?: number | null;
  movement_speed_mph?: number | null;
  last_update?: string | null;
  public_advisory?: StormAdvisoryLink | null;
  forecast_advisory?: StormAdvisoryLink | null;
  forecast_discussion?: StormAdvisoryLink | null;
  forecast_graphics?: StormAdvisoryLink | null;
  forecast_track_kmz_url?: string | null;
  wind_speed_probabilities_url?: string | null;
}

export interface StormTrackingSnapshot {
  source_url: string;
  status: "available" | "unavailable";
  fetched_at?: string | null;
  message?: string | null;
  active_storms: StormSystem[];
}
